src/controller/CHIPDeviceController.cpp

```diff
@@ -1178,6 +1184,298 @@ void DeviceCommissioner::OnSessionEstablished()
+    ByteSpan vendorReserved1;
+    ByteSpan vendorReserved2;
+    ByteSpan vendorReserved3;
+    ByteSpan vendorReserved4;
+
+    ByteSpan * element_array[] = { &certDeclaration, &attestationNonce, &timestamp,       &firmwareInfo,
+                                   &vendorReserved1, &vendorReserved2,  &vendorReserved3, &vendorReserved4 };
+
+    System::PacketBufferTLVWriter tlvWriter;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+    System::PacketBufferHandle attestationElementsTbs;
+
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+    int i                      = 0;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+    do
+    {
+        if (!TLV::IsContextTag(tlvReader.GetTag()))
+        {
+            continue;
+        }
+        const uint8_t * data = nullptr;
+        ReturnErrorOnFailure(tlvReader.GetDataPtr(data));
+        *element_array[i++] = ByteSpan(data, tlvReader.GetLength());
+    } while (tlvReader.Next() == CHIP_NO_ERROR);
+
+    // Step f
+    // Retrieve attestation challenge
+    PeerConnectionState * state = mSessionMgr->GetPeerConnectionState(
+        { mPairingSession.PeerConnection().GetPeerNodeId(), mPairingSession.GetPeerKeyId(), mAdminId });
+    VerifyOrReturnError(state != nullptr, CHIP_ERROR_NOT_CONNECTED);
+    ByteSpan attestationChallenge = state->GetSecureSession().GetAttestationChallenge();
+
+    // Create attestation-elements-tbs
+    attestationElementsTbs = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    VerifyOrReturnError(!attestationElementsTbs.IsNull(), CHIP_ERROR_NO_MEMORY);
+
+    tlvWriter.Init(std::move(attestationElementsTbs));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(0), attestationElements));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(1), attestationChallenge));
+    ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Finalize(&attestationElementsTbs));
+
+    // Step e
+    P256ECDSASignature deviceSignature;
+    ReturnErrorOnFailure(deviceSignature.SetLength(signature.size()));
+    memcpy(deviceSignature, signature.data(), signature.size());
+    ReturnErrorOnFailure(mRemoteManufacturerPubkey.ECDSA_validate_msg_signature(
+        attestationElementsTbs->Start(), attestationElementsTbs->DataLength(), deviceSignature));
+
+    // Step g
+    VerifyOrReturnError(sizeof(mAttestationNonce) == attestationNonce.size(), CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(memcmp(mAttestationNonce, attestationNonce.data(), sizeof(mAttestationNonce)) == 0, CHIP_ERROR_INTERNAL);
```

Comment:
This needs to be validate before the signature is checked.

Reply:
done